### DISCUSSION TITLE
how to create conditional cascades ?

### DISCUSSION THREAD

**Discussion by johaven** [sqlalchemy/sqlalchemy#9923] [Usage Questions] [answered]
```python
class Comment(Base):
    __tablename__ = 'comments'
    __table_args__ = Base.insensitive
    id = db.Column(db.BigInteger, primary_key=True)
    uid = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), index=True, nullable=True)
    guid = db.Column(db.Integer, db.ForeignKey('guests.id', ondelete='CASCADE'), index=True, nullable=True)
    sid = db.Column(db.BigInteger, db.ForeignKey('shares.id', ondelete='CASCADE'), index=True, nullable=True)
    fid = db.Column(db.BigInteger, db.ForeignKey('files.id', ondelete='CASCADE'), index=True, nullable=True)
```

I would like to create conditional cascades but i don't know how to do it.
If the _fid_ and _sid_ are not null and the share (_sid_) is deleted, I would like to set the sid value to null instead of deleting the comment.
In the other cases if the fid or the sid are present individually on the comment, the behavior of the current cascade suits me.

**Comment by CaselIT** [comment, marked as answer]
Hi,

The ondelete and onupdate options of a forign key are passed directly to the db when creating the table, so they don't control how sqlalchemy behaves.

To implement a logic like what you describe you could probably use the event system, to intercept deletion of a `sid` and update the comments connected to it setting the value to null.
This event could probably work https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_delete
